cou/steps/plan.py

```diff
@@ -317,6 +318,75 @@ def _get_os_release_and_series(analysis_result: Analysis) -> tuple[OpenStackRele
     return current_os_release, current_series
 
 
+def _generate_control_plane_plan(
+    target: OpenStackRelease, apps: list[OpenStackApplication], force: bool
+) -> list[UpgradePlan]:
+    """Generate upgrade plan for control plane.
+
+    :param target: Target OpenStack release.
+    :type target: OpenStackRelease
+    :param apps: List of control plane applications.
+    :type apps: list[OpenStackApplication]
+    :param force: Whether the plan generation should be forced
+    :type force: bool
+    :return: Principal and Subordiante upgrade plan.
+    :rtype: list[UpgradePlan]
+    """
+    principal_upgrade_plan = create_upgrade_group(
+        apps=apps,
+        description="Control Plane principal(s) upgrade plan",
+        target=target,
+        force=force,
+        filter_function=lambda app: not isinstance(app, SubordinateBaseClass),
+    )
+
+    subordinate_upgrade_plan = create_upgrade_group(
+        apps=apps,
+        description="Control Plane subordinate(s) upgrade plan",
+        target=target,
+        force=force,
+        filter_function=lambda app: isinstance(app, SubordinateBaseClass),
+    )
+
+    logger.debug("generating of control plan upgrade plan finished")
+    return [principal_upgrade_plan, subordinate_upgrade_plan]
+
+
+def _generate_data_plane_plan(
+    target: OpenStackRelease, apps: list[OpenStackApplication], force: bool
+) -> list[UpgradePlan]:  # pragma: no cover
+    """Generate upgrade plan for data plane.
+
+    :param target: Target OpenStack release.
+    :type target: OpenStackRelease
+    :param apps: List of data plane applications.
+    :type apps: list[OpenStackApplication]
+    :param force: Whether the plan generation should be forced
+    :type force: bool
+    :return: Hypervisors plans and other
+    :rtype: list[UpgradePlan]
+    """
+    nova_compute_machines = [
+        unit.machine.machine_id
+        for app in apps
+        for unit in app.units.values()
+        if app.charm == "nova-compute"
+    ]
+    hypervisor_apps = []
+    for app in apps:
+        for machine in app.machines:
+            if machine in nova_compute_machines:
+                hypervisor_apps.append(app)
+                break  # exiting machine for loop, since
+
+    hypervisor_planner = HypervisorUpgradePlanner(hypervisor_apps)
+    hypervisor_plans = hypervisor_planner.generate_upgrade_plan(target, force)
+    logger.debug("generating of hypervisors upgrade plan finished")
+
+    # generating rest
```

Comment:
didn't understand this comment

Reply:
For example, the ceph-osd will be there.

Reply:
```suggestion
    # generate the plan for the remaining data-plane apps
```
See my comment above of removing ceph-osd from hypervisor_apps.

Reply:
You know that this comment is just for you now and will be removed after we actually do ceph-osd.